packages/cli/src/modules/insights/__tests__/insights.service.test.ts

```diff
@@ -674,3 +674,281 @@ describe('getInsightsSummary', () => {
+		// ARRANGE
+		for (const workflow of [workflow1, workflow2, workflow3]) {
+			await createCompactedInsightsEvent(workflow, {
+				type: 'success',
+				value: workflow === workflow1 ? 1 : workflow === workflow2 ? 2 : 3,
+				periodUnit: 'day',
+				periodStart: DateTime.utc(),
+			});
+		}
+
+		// ACT
+		const byWorkflow = await insightsService.getInsightsByWorkflow({
+			nbDays: 14,
+			sortBy: 'succeeded:desc',
+			skip: 1,
+			take: 1,
+		});
+
+		// ASSERT
+		expect(byWorkflow.count).toEqual(3);
+		expect(byWorkflow.data).toHaveLength(1);
+		expect(byWorkflow.data[0].workflowId).toEqual(workflow2.id);
+	});
+
+	test('compacted data are grouped by workflow correctly even with 0 data (check division by 0)', async () => {
+		// ACT
+		const byWorkflow = await insightsService.getInsightsByWorkflow({
+			nbDays: 14,
+		});
+
+		// ASSERT
+		expect(byWorkflow.count).toEqual(0);
+		expect(byWorkflow.data).toHaveLength(0);
+	});
+});
+
+describe('getInsightsByTime', () => {
+	let insightsService: InsightsService;
+	beforeAll(async () => {
+		insightsService = Container.get(InsightsService);
+	});
+
+	let project: Project;
+	let workflow1: IWorkflowDb & WorkflowEntity;
+	let workflow2: IWorkflowDb & WorkflowEntity;
+
+	beforeEach(async () => {
+		await truncateAll();
+
+		project = await createTeamProject();
+		workflow1 = await createWorkflow({}, project);
+		workflow2 = await createWorkflow({}, project);
+	});
+
+	test('returns empty array when no insights exist', async () => {
+		const byTime = await insightsService.getInsightsByTime(14);
+		expect(byTime).toEqual([]);
+	});
+
+	test('returns empty array when no insights in the time range exists', async () => {
+		await createCompactedInsightsEvent(workflow1, {
+			type: 'success',
+			value: 2,
+			periodUnit: 'day',
+			periodStart: DateTime.utc().minus({ days: 30 }),
+		});
+
+		const byTime = await insightsService.getInsightsByTime(14);
+		expect(byTime).toEqual([]);
+	});
+
+	test('compacted data are are grouped by time correctly', async () => {
+		// ARRANGE
+		for (const workflow of [workflow1, workflow2]) {
+			await createCompactedInsightsEvent(workflow, {
+				type: 'success',
+				value: workflow === workflow1 ? 1 : 2,
+				periodUnit: 'day',
+				periodStart: DateTime.utc(),
+			});
+			await createCompactedInsightsEvent(workflow, {
+				type: 'failure',
+				value: 2,
+				periodUnit: 'day',
+				periodStart: DateTime.utc(),
+			});
+			await createCompactedInsightsEvent(workflow, {
+				type: 'success',
+				value: 1,
+				periodUnit: 'day',
+				periodStart: DateTime.utc().minus({ day: 2 }),
+			});
+			await createCompactedInsightsEvent(workflow, {
+				type: 'success',
+				value: 1,
+				periodUnit: 'day',
+				periodStart: DateTime.utc().minus({ days: 10 }),
+			});
+			await createCompactedInsightsEvent(workflow, {
+				type: 'runtime_ms',
+				value: workflow === workflow1 ? 10 : 20,
+				periodUnit: 'day',
+				periodStart: DateTime.utc().minus({ days: 10 }),
+			});
```

Comment:
Could we add events that go across the edge of the filter, e.g. an event that was  14 days ago and one that was 13 days ago.

May take a bit of massaging since we cut off at utc midnight and not `nbDays * 24` hours ago.

Reply:
Same for the test of the other method.

Reply:
I added some tests here: https://github.com/n8n-io/n8n/pull/14064/files#diff-558eb3d4d004e6e0f0244a529b934ce0197937549d125bebe56a58c49acd94dbR908 (and same for workflow by time)

Reply:
Nice, I'll add another insight that is just barely in the cut off period.

Reply:
Doing this for `getInsightsByWorkflow` I realized that it treats dates differently compared to `getInsightsByTime`. 
`getInsightsByTime` is using `date`, `CURRENT_DATE` and `CURDATE`, whereas `getInsightsByWorkflow` uses `datetime` and `NOW` respectively. 

This means `getInsightsByTime` will cut off the period at midnight UTC, and `getInsightsByWorkflow` will cut off at the current time minus `maxAgeInDays`.

Is this intentional?

Reply:
Well, here were my thoughts about it in a diagram:
- the 7 day sliding window made sense in the scope of a global overview (its only one number that reflect the last 7 days)
- the 7 day sliding window did not make sense for a bar chart, because the oldest day would be cut off, and could not compare to the other ones

Here is the diagram: https://excalidraw.com/#json=mrVpLXmeewjALIS-S-05-,CoyOqv7PL7AV9HA8UETapA
So in my mind, it made sense to have different time windows for overview and insights, but I'm not sure

Reply:
Hey @despairblue I implement the 7 day rolling window strategy (we accept the "dip" for oldest day in graphs)